### PR TITLE
Add Cylinder function

### PR DESCRIPTION
**Pull request by TiagoCavalcanteTrindade**
Rebase of #1450.

Closes #1294.

### CODE REVIEW

**Comment by TiagoCavalcanteTrindade** [comment]
@rocky could you review my changes? I put a FIXME [here](https://github.com/TiagoCavalcanteTrindade/Mathics/blob/add-cylinder-function/mathics/format/asy.py#L453) I don't know to transform the current data in a *draw around this axis*.

File: mathics/format/asy.py [hunk 1]
```diff
@@ -310,7 +311,7 @@ def polygon3dbox(self, **options) -> str:
 add_conversion_fn(Polygon3DBox)
 
 
-def polygonbox(self, **options):
+def polygonbox(self, **options) -> str:
```
Comment (rocky) [inline mathics/format/asy.py:314]: Routines to draw SVG are no longer here but are in `format/svg.py`

Comment (TiagoCavalcanteTrindade) [reply mathics/format/asy.py:314]: Oh, ok I gonna rebase it.

File: mathics/format/asy.py [hunk 2]
```diff
@@ -427,3 +428,34 @@ def sphere3dbox(self, **options) -> str:
 
 
 add_conversion_fn(Sphere3DBox)
+
+
+def cylinder3dbox(self, **options) -> str:
```
Comment (rocky) [inline mathics/format/asy.py:433]: Also move  - here into `format/asy.py`

File: mathics/format/asy.py (hunk 1 above)
Comment (rocky) [reply mathics/format/asy.py:314]: rebase doesn't track refactoring like this.  You will probably have to move it by hand.

**Comment by rocky** [comment]
> @rocky could you review my changes? I put a FIXME [here](https://github.com/TiagoCavalcanteTrindade/Mathics/blob/add-cylinder-function/mathics/format/asy.py#L453) I don't know to transform the current data in a _draw around this axis_.

Right now there isn't a transform function, but there should be one. In #1373 there is code that needs to be adapted [here](https://github.com/mathics/Mathics/blob/0a283aa975225d2c6ed04b09feab3d219ec4774c/mathics/builtin/tensors.py#L498-L583)

**Comment by TiagoCavalcanteTrindade** [comment]
@rocky you think it's better we do that before we merge?

Also, what do you think about the [odd number of coordinates handling](https://github.com/Mathics3/mathics-django/pull/31#issuecomment-868875962)?

**Comment by rocky** [comment]
Here is a different idea for how to approach this and at the same time go in the direction we want to go in....

Write a `to_json()` routine for this. But actually, I'd start out with sphere, since that might be easier and here we want to focus on the JSON aspect. 

Then we should start a graphics-backend module. (In the past, it was erroneously called a pymathics module, but I realize this was wrong. This is a pure module  _of any kind_ that takes JSON with our simple graphics API and converts it into threejs.   You could start out with `graphics3d.js` and/or `mathics.js` from the Django repository and turn that into a nodejs module (or however you want to package it).

**Comment by rocky** [comment]
> @rocky you think it's better we do that before we merge?
> 
> Also, what do you think about the [odd number of coordinates handling](https://github.com/Mathics3/mathics-django/pull/31#issuecomment-868875962)?

Release 3.0.0 of Mathics core and Django are currently _out_. Look on PyPI or in "releases" of this project.

**Comment by rocky** [comment]
> Release 3.0.0 of Mathics core and Django are currently _out_. Look on PyPI or in "releases" of this project.

I am still working on mathics omnibus though.

**Comment by TiagoCavalcanteTrindade** [comment]
> > @rocky you think it's better we do that before we merge?
> > Also, what do you think about the [odd number of coordinates handling](https://github.com/Mathics3/mathics-django/pull/31#issuecomment-868875962)?
> 
> Release 3.0.0 of Mathics core and Django are currently _out_. Look on PyPI or in "releases" of this project.

@rocky I didn't find the information I need here.

> Write a to_json() routine for this. But actually, I'd start out with sphere, since that might be easier and here we want to focus on the JSON aspect.

Ok, I guess it should be in another PR, isn't it?

> Then we should start a graphics-backend module. (In the past, it was erroneously called a pymathics module, but I realize this was wrong. This is a pure module of any kind that takes JSON with our simple graphics API and converts it into threejs. You could start out with graphics3d.js and/or mathics.js from the Django repository and turn that into a nodejs module (or however you want to package it).

@rocky if you are talking about keeping the transformations in 2 places I think there's no problem, but otherwise  I'm afraid it has too many layers in the code.

**Comment by rocky** [comment]
> > Release 3.0.0 of Mathics core and Django are currently _out_. Look on PyPI or in "releases" of this project.
> 
> @rocky I didn't find the information I need here.

I don't understand what information you need. 

All I was saying is that there are new releases: https://pypi.org/project/Mathics3/ (pypi) and https://github.com/mathics/Mathics/releases/tag/3.0.0

What 
> 
> > Write a to_json() routine for this. But actually, I'd start out with sphere, since that might be easier and here we want to focus on the JSON aspect.
> 
> Ok, I guess it should be in another PR, isn't it?
> 

Yes, all of this is a new PR. New approach, new PR. And let's try to start simple to work up to cylinders which add this additional wrinkle.

**Comment by TiagoCavalcanteTrindade** [comment]
> I don't understand what information you need.

Sorry for the confusion.

The information I was needing was what we should do when the user does something like:
```wl
Cylinder[{{0,0,0}}] (* it should raise an error because of the user didn't provided an end point, just the start point *)
```

But you already answered [here](https://github.com/Mathics3/mathics-django/pull/31#issuecomment-869073001).

**Comment by TiagoCavalcanteTrindade** [comment]
I left a FIXME saying we need to do the following `transformation_vector * (radius, radius, radius)` instead of `(1, 1, 1) * (radius, radius, radius)`. The current version does the plot range be bigger than it should.

As we are going to do the transformation part in other PR I guess this should be merged before that.

Also, the following is the behavior of an odd number of points: 
image

**Review by rocky — APPROVED** [review]
LGTM - I wonder why we get the spurious black style wars. 

Did you run `black .` ?

**Comment by TiagoCavalcanteTrindade** [comment]
@rocky no, I'm doing that (I've forgotten).

**Comment by rocky** [comment]
I suspect then it is the black version that is used. What does `black --version` say?

**Comment by TiagoCavalcanteTrindade** [comment]
@rocky I don't have git pre-commit installed, I guess it's that.

My black version is 21.6b0.

**Comment by rocky** [comment]
Ok - Let's set everything to 21.6b0. 

I will add a check to the workflows too.

**Pull request merged**